cudax/include/cuda/experimental/__async/sender/completion_signatures.cuh

```diff
@@ -24,24 +24,418 @@
 #include <cuda/std/__type_traits/conditional.h>
 #include <cuda/std/__type_traits/disjunction.h>
 #include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/is_base_of.h>
+#include <cuda/std/__type_traits/is_callable.h>
+#include <cuda/std/__type_traits/type_list.h>
+#include <cuda/std/__utility/typeid.h>
+#include <cuda/std/tuple>
 
 #include <cuda/experimental/__async/sender/cpos.cuh>
 #include <cuda/experimental/__async/sender/exception.cuh>
+#include <cuda/experimental/__async/sender/tuple.cuh>
 #include <cuda/experimental/__async/sender/type_traits.cuh>
+#include <cuda/experimental/__detail/config.cuh>
 
+#include <nv/target>
+
+// include this last:
 #include <cuda/experimental/__async/sender/prologue.cuh>
 
 namespace cuda::experimental::__async
 {
+template <class... _Ts>
+using __ts = _CUDA_VSTD::__type_list<_Ts...>;
+
+template <class _ValueTuplesList = _CUDA_VSTD::__type_list<>,
+          class _ErrorsList      = _CUDA_VSTD::__type_list<>,
+          bool _HasStopped       = false>
+struct __partitioned_completions;
+
+template <class _Tag>
+struct __partitioned_fold_fn;
+
+// The following overload of binary operator* is used to build up the cache of
+// completion signatures. We fold over operator*, accumulating the completion
+// signatures in the cache. `__undefined` is used here to prevent the
+// instantiation of the intermediate types.
+template <class _Partitioned, class _Tag, class... _Args>
+_CUDAX_API auto operator*(__undefined<_Partitioned>&, _Tag (*)(_Args...))
+  -> _CUDA_VSTD::__call_result_t<__partitioned_fold_fn<_Tag>, _Partitioned&, __undefined<__ts<_Args...>>&>;
+
+// This unary overload is used to extract the cache from the `__undefined` type.
+template <class _Partitioned>
+_CUDAX_API auto __unpack_partitioned_completions(__undefined<_Partitioned>&) -> _Partitioned;
+
+template <class... _Sigs>
+using __partition_completion_signatures_t _CCCL_NODEBUG_ALIAS = //
+  decltype(__async::__unpack_partitioned_completions(
+    (declval<__undefined<__partitioned_completions<>>&>() * ... * static_cast<_Sigs*>(nullptr))));
```

Comment:
Nitpick: I *think* this is safe to leave unqualified because we cannot pull in any other associated namespace through template arguments, but I am a big fan of `always qualify` if you do not want to use ADL